### PR TITLE
Porting rosbridge library (capabilities and internal) to ROS2.

### PR DESCRIPTION
**Pull request by jubeira**
See https://github.com/ros2/rosbridge_suite/issues/3.

With this patch, all the code from the entrypoint (`Websocket server`) up to the internal libraries should be ROS2-compatible, and no references to `rclpy` or ROS1 code should remain.

<details><summary>Details</summary>

- All the ROS1 code that involves publishers, subscribers and services now need a `rclpy.node.Node` handle to work.
- `RosbridgeWebsocketNode` in `rosbridge_websocket` module was converted to a ROS2 node because it's near the entry point of the application. A reference to it shall be used to generate publishers, subscribers and service clients where necessary.
- Besides routing the node handle to all the points of the program where it was required (the deepest point from the entrypoint is the `internal` package), there were a few features from ROS1 that changed a bit, so workarounds had to be found:
  - In `publishers` module inside `internal`, a ROS1 `SubscriberListener` was used to prevent late-joining subscribers from missing messages. This is now achieved using transient local QoS with a lifespan.
  - At the same time, there was support for latched topics. This now applies for publishers and subscribers, so transient local is the QoS of choice for subscribers as well.
  - In `subscribers` module, additional subscriber callbacks were used to send latched messages to new (web) subscribers using only one ROS subscriber. Since this is not available in ROS2 (at least not that I know of), a workaround was implemented and tested in a minimal example. 
- In `message_conversion`, the parameters don't seem to be used in the original code. For a first iteration, using defaults should work.
</details>

### CODE REVIEW

**Comment by jubeira** [comment]
@ivanpauno I can't add you as a reviewer, but would you please take an initial look?

/cc @dirk-thomas

File: rosbridge_server/scripts/rosbridge_websocket.py [hunk 1]
```diff
@@ -55,184 +60,240 @@
...
+class RosbridgeWebsocketNode(Node):
+    def __init__(self):
+        super().__init__('rosbridge_websocket')
+
+        ##################################################
+        # Parameter handling                             #
+        ##################################################
+        retry_startup_delay = self.get_parameter('retry_startup_delay').value
+        retry_startup_delay = 2.0 if retry_startup_delay is None
+                                  else retry_startup_delay  # seconds
```
Comment (ivanpauno) [inline rosbridge_server/scripts/rosbridge_websocket.py]: I think you can use `get_parameter_or` to avoid this logic, right?

Comment (ivanpauno) [reply rosbridge_server/scripts/rosbridge_websocket.py]: same in many of the other calls to `get_parameter` here

File: rosbridge_server/scripts/rosbridge_websocket.py [hunk 2]
```diff
@@ -55,184 +60,240 @@
...
+
+        bson_only_mode = self.get_parameter('retry_startup_delay').value
+        bson_only_mode = False if bson_only_mode is None
+                               else bson_only_mode
+
+        if RosbridgeWebSocket.max_message_size == "None":
+            RosbridgeWebSocket.max_message_size = None
+
+        # SSL options
+        certfile = self.get_parameter('certfile').value
```
Comment (ivanpauno) [inline rosbridge_server/scripts/rosbridge_websocket.py]: I think this can raise an error if you don't allow undeclared parameters. I would allow them, or use `get_paramer_or` again here.
I think you will need `automatically_declare_parameter_overrides` too, right?

File: rosbridge_server/scripts/rosbridge_websocket.py [hunk 3]
```diff
@@ -55,184 +60,240 @@
...
+
+        port = self.get_parameter('port').value
+        port = 9090 if port is None else port
+        
+        address = self.get_parameter('address').value
+        address = "" if address is None else address
+
+        # Publisher for number of connected clients
+        # QoS profile with transient local durability (latched topic in ROS 1).
+        client_count_qos_profile = qos_profile_default
```
Comment (ivanpauno) [inline rosbridge_server/scripts/rosbridge_websocket.py]: I think `qos_profile_default` is deprecated. Also, why not setting `depth` to 10? `queue_size` was 10 before.

File: rosbridge_server/scripts/rosbridge_websocket.py [hunk 4]
```diff
@@ -55,184 +60,240 @@
...
+
+        RosbridgeWebSocket.client_count_pub = self.create_publisher(Int32, 'client_count',
+                                                                    qos_profile=client_count_qos_profile)
+        RosbridgeWebSocket.client_count_pub.publish(Int32(data=0))
+
+        # Get the glob strings and parse them as arrays.
+        topics_glob = self.get_parameter('topics_glob').value
+        topics_glob = '' if topics_glob is None else topics_glob
+
+        services_glob = self.get_parameter('topics_glob').value
```
Comment (ivanpauno) [inline rosbridge_server/scripts/rosbridge_websocket.py]: topics -> services

File: rosbridge_server/scripts/rosbridge_websocket.py [hunk 5]
```diff
@@ -55,184 +60,240 @@
...
+        RosbridgeWebSocket.services_glob = [
+            element.strip().strip("'")
+            for element in services_glob[1:-1].split(',')
+            if len(element.strip().strip("'")) > 0]
+        RosbridgeWebSocket.params_glob = [
+            element.strip().strip("'")
+            for element in params_glob[1:-1].split(',')
+            if len(element.strip().strip("'")) > 0]
+
+        if "--port" in sys.argv:
```
Comment (ivanpauno) [inline rosbridge_server/scripts/rosbridge_websocket.py]: I would consider start using `argparse`.

Comment (ivanpauno) [reply rosbridge_server/scripts/rosbridge_websocket.py]: Maybe, that can be done in the original code too.

File: rosbridge_library/src/rosbridge_library/internal/services.py [hunk 6]
```diff
@@ -94,26 +94,34 @@ def args_to_service_request_instance(service, inst, args):
...
     args_to_service_request_instance(service, inst, args)
 
-    # Call the service
-    proxy = ServiceProxy(service, service_class)
-    response = proxy.call(inst)
+    client = node_handle.create_client(service_class, service)
 
-    # Turn the response into JSON and pass to the callback
-    json_response = extract_values(response)
+    future = client.call_async(inst)
```
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rosbridge_library/src/rosbridge_library/internal/services.py]: Maybe, `call_async` usage is fine.

File: rosbridge_library/src/rosbridge_library/internal/message_conversion.py [hunk 7]
```diff
@@ -82,20 +81,33 @@
                                 ("char[]", re.compile(r'char\[[^\]]*\]'))]
 
 binary_encoder = None
+binary_encoder_type = 'default'
+bson_only_mode = False
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/message_conversion.py]: Yes; the problem here is that it's a bit difficult to get a node handle to retrieve parameters at this point in the code as the `TODO` below indicates.

In any case `binary_encoder_type` is an argument to the launchfiles but it's not used anywhere as a ROS parameter (see e.g. [websocket server](https://github.com/RobotWebTools/rosbridge_suite/blob/0.11.2/rosbridge_server/launch/rosbridge_websocket.launch#L25).

`bson_only_mode` on the other hand [is used](https://github.com/RobotWebTools/rosbridge_suite/blob/0.11.2/rosbridge_server/launch/rosbridge_websocket.launch#L62), but for a first iteration we can start with a default value and then improve the mechanism to use the parameter.

[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/message_conversion.py]: Yes; the problem here is that it's a bit difficult to get a node handle to retrieve parameters at this point in the code as the `TODO` below indicates.

In any case `binary_encoder_type` is an argument to the launchfiles but it's not used anywhere as a ROS parameter (see e.g. [websocket server](https://github.com/RobotWebTools/rosbridge_suite/blob/0.11.2/rosbridge_server/launch/rosbridge_websocket.launch#L25).

`bson_only_mode` on the other hand [is used](https://github.com/RobotWebTools/rosbridge_suite/blob/0.11.2/rosbridge_server/launch/rosbridge_websocket.launch#L62), but for a first iteration we can start with a default value and then improve the mechanism to use the parameter.

File: rosbridge_library/src/rosbridge_library/internal/publishers.py [hunk 8]
```diff
@@ -256,7 +195,7 @@ def __init__(self):
         self.unregister_timers = {}
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/publishers.py]: That's a good question indeed.
`PublisherManager` (and the analogous manager for subscribers) is a class that is defined here and is only used within this module to create a global manager, which is then imported at some other points of the code (see [here](https://github.com/RobotWebTools/rosbridge_suite/pull/419/files/0fffe611df42bc37b2a90e8a7d62670226a4590c#diff-5de390a84a194f6692ef968c0c177283L359)).

In other words, these managers are some sort of singletons which are used throughout the code, and therefore there is no way of sending a node handle as a parameter upon construction. On the other hand, the handle is available when the methods that actually need it are called, and that's why it is the selected approach.

File: rosbridge_library/src/rosbridge_library/internal/subscribers.py [hunk 9]
```diff
@@ -193,12 +224,14 @@ def subscribe(self, client_id, topic, callback, msg_type=None):
...
-            self._subscribers[topic] = MultiSubscriber(topic, msg_type)
+            self._subscribers[topic] = MultiSubscriber(
+                topic, client_id, callback, node_handle, msg_type=msg_type)
+        else:
+            self._subscribers[topic].subscribe(client_id, callback)
 
         if msg_type is not None:
             self._subscribers[topic].verify_type(msg_type)
 
-        self._subscribers[topic].subscribe(client_id, callback)
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/subscribers.py]: In the old code this line triggered the 'hack' of adding and removing a callback, receiving a new message if the topic is latched. This happened with every call of this method, regardless if the `MultiSubscriber` was created in this call or not.

The patch proposed for latched topics is creating a new subscriber that processes the first message, updates the subscriptions and then is removed. But you don't need to do this with the first client id when `MultiSubscriber` is created, since the subscription will get any latched messages. That's why the new constructor initializes itself with a client id, and avoids creating a second subscription that uses `_new_sub_callback`.

I'm inclined for this approach as it is more efficient. Creating a `MultiSubscriber` without an initial client id has no use, as it has to be added afterwards anyway. Creating it with a client id avoids the need of creating a second subscription with a different callback just to end up using the first subscription originally created in the constructor.

File: rosbridge_server/scripts/rosbridge_websocket.py (hunk 1 above)
Comment (jubeira) [reply rosbridge_server/scripts/rosbridge_websocket.py]: Yes; that wasn't part of `rclpy` back then when I made these changes.
+1 to changing it to `get_parameter_or`.

File: rosbridge_library/src/rosbridge_library/internal/subscribers.py [hunk 10]
```diff
@@ -69,12 +70,15 @@ def __init__(self, topic, msg_type=None):
 
         """
         # First check to see if the topic is already established
-        topic_type = get_topic_type(topic)[0]
+        topics_names_and_types = dict(node_handle.get_topic_names_and_types())
+        topic_type = topics_names_and_types.get(topic)
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/subscribers.py]: I think it should be fine as is; thanks for raising this anyways.

File: rosbridge_library/src/rosbridge_library/internal/subscribers.py [hunk 11]
```diff
@@ -43,19 +41,22 @@
...
-
 class MultiSubscriber():
     """ Handles multiple clients for a single subscriber.
 
     Converts msgs to JSON before handing them to callbacks.  Due to subscriber
     callbacks being called in separate threads, must lock whenever modifying
     or accessing the subscribed clients. """
 
-    def __init__(self, topic, msg_type=None):
+    def __init__(self, topic, client_id, callback, node_handle, msg_type=None):
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/subscribers.py]: See https://github.com/RobotWebTools/rosbridge_suite/pull/419#discussion_r305935801.

File: rosbridge_library/src/rosbridge_library/internal/subscribers.py [hunk 12]
```diff
@@ -125,11 +134,15 @@ def subscribe(self, client_id, callback):
...
             # the given callback.
-            self.subscriber.impl.add_callback(self.callback, [callback])
-            self.subscriber.impl.remove_callback(self.callback, [callback])
+            # In any case, the first message is handled using new_sub_callback,
+            # which adds the new callback to the subscriptions dictionary.
+            qos_profile = QoSProfile(durability=QoSDurabilityPolicy.RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL)
+            self.new_subscriptions.update({client_id: callback})
+            if self.new_subscriber is None:
+                self.new_subscriber = self.node_handle.create_subscription(
+                    self.msg_class, self.topic, self._new_sub_callback, qos_profile=qos_profile)
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/subscribers.py]: It is indeed a bit of a workaround; the point is to get a (potentially) 'latched' message from a topic with an existing subscriber.

Using one subscriber per client id could work as well for that case, but in the case the topic not latched and a publisher is sending data at a high rate it would be worse than this solution IMO.

File: rosbridge_server/scripts/rosbridge_websocket.py (hunk 5 above)
Comment (jubeira) [reply rosbridge_server/scripts/rosbridge_websocket.py]: Yes, this seems to be like a manually implemented `argparse`. I'd say it's out of scope of this PR.

File: rosbridge_library/src/rosbridge_library/internal/message_conversion.py [hunk 13]
```diff
@@ -221,14 +239,15 @@ def _to_binary_inst(msg):
 
 def _to_time_inst(msg, rostype, inst=None):
     # Create an instance if we haven't been provided with one
+
     if rostype == "time" and msg == "now":
-        return rospy.get_rostime()
+        return ROSClock().now().to_msg()
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/message_conversion.py]: Thanks; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_library/src/rosbridge_library/internal/services.py [hunk 14]
```diff
@@ -94,26 +94,34 @@ def args_to_service_request_instance(service, inst, args):
...
 
 
-def call_service(service, args=None):
+def call_service(node_handle, service, args=None):
     # Given the service name, fetch the type and class of the service,
     # and a request instance
 
-    service = resolve_name(service)
+    # TODO(@jubeira): check if service name requires processing
+    # (equivalent to rospy.resolve_name)
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/services.py]: Thanks; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_library/src/rosbridge_library/internal/services.py [hunk 15]
```diff
@@ -94,26 +94,34 @@ def args_to_service_request_instance(service, inst, args):
...
 
-    service_type = get_service_type(str(service))
+    service_names_and_types = dict(node_handle.get_service_names_and_types())
+    service_type = service_names_and_types.get(service)
     if service_type is None:
         raise InvalidServiceException(service)
+    # service_type is a tuple of types at this point; only one type is supported.
+    service_type = service_type[0]
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/services.py]: Thanks; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_library/src/rosbridge_library/internal/subscribers.py [hunk 16]
```diff
@@ -69,12 +70,15 @@ def __init__(self, topic, msg_type=None):
...
-        topic_type = get_topic_type(topic)[0]
+        topics_names_and_types = dict(node_handle.get_topic_names_and_types())
+        topic_type = topics_names_and_types.get(topic)
 
         # If it's not established and no type was specified, exception
         if msg_type is None and topic_type is None:
             raise TopicNotEstablishedException(topic)
 
+        # topic_type is a list of types at this point; only one type is supported.
+        topic_type = topic_type[0]
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/subscribers.py]: Thanks; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_library/src/rosbridge_library/internal/subscribers.py [hunk 17]
```diff
@@ -170,9 +183,27 @@ def callback(self, msg, callbacks=None):
...
+        If the topic was latched, a new subscriber has to be added to receive
+        a new message and route it to the new subscriptor.
+
+        After the first message is routed, the new subscriber is deleted and
+        the subscriptions dictionary is updated with the newly incorporated
+        subscriptors.
+        """
+        with self.lock:
+            self.callback(msg, [self.new_subscriptions.values()])
+            self.subscriptions.update(new_subscriptions)
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/internal/subscribers.py]: Good catch; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_library/src/rosbridge_library/rosbridge_protocol.py [hunk 18]
```diff
@@ -55,8 +55,8 @@ class RosbridgeProtocol(Protocol):
 
     parameters = None
 
-    def __init__(self, client_id, parameters = None):
+    def __init__(self, client_id, parameters = None, node_handle = None):
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_library/src/rosbridge_library/rosbridge_protocol.py]: It shouldn't be an optional argument indeed. Addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_server/src/rosbridge_server/websocket_handler.py [hunk 19]
```diff
@@ -80,6 +79,8 @@ class RosbridgeWebSocket(WebSocketHandler):
     max_message_size = None                 # bytes
     unregister_timeout = 10.0               # seconds
     bson_only_mode = False
+    _node_handle = None
```
[the remark this answers is not in the document]
Comment (jubeira) [reply rosbridge_server/src/rosbridge_server/websocket_handler.py]: Good catch; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_server/scripts/rosbridge_websocket.py (hunk 1 above)
Comment (jubeira) [reply rosbridge_server/scripts/rosbridge_websocket.py]: Changed to `declare_parameter` in Thanks; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_server/scripts/rosbridge_websocket.py (hunk 2 above)
Comment (jubeira) [reply rosbridge_server/scripts/rosbridge_websocket.py]: Changed to `declare_parameter` in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_server/scripts/rosbridge_websocket.py (hunk 4 above)
Comment (jubeira) [reply rosbridge_server/scripts/rosbridge_websocket.py]: Thanks; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

File: rosbridge_server/scripts/rosbridge_websocket.py (hunk 3 above)
Comment (jubeira) [reply rosbridge_server/scripts/rosbridge_websocket.py]: Thanks; addressed in https://github.com/RobotWebTools/rosbridge_suite/pull/419/commits/9161521da0b402b5fa6720e54cb61bd156ec58f6.

**Comment by jubeira** [comment]
This was rebased against `develop`; the latest changes were ported to ROS2 as well.

Initial E2E tests are passing for topics, services and parameters using standard `roslibjs`. This is ready for another review.

**Comment by jubeira** [comment]
@gonzodepedro would you please take one more look? Thanks!

**Comment by jubeira** [comment]
Thanks @gonzodepedro !
